crates/ntx-builder/src/builder/data_store/account_cache.rs

```diff
@@ -0,0 +1,179 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    sync::Mutex,
+};
+
+use miden_objects::{
+    account::{Account, AccountId},
+    note::{NoteExecutionMode, NoteTag},
+};
+
+// ACCOUNT CACHE
+// =================================================================================================
+
+#[derive(Debug, Hash, Clone, Copy, PartialEq, Eq)]
+pub struct NtxAccountIdPrefix(u32);
+
+impl From<AccountId> for NtxAccountIdPrefix {
+    fn from(id: AccountId) -> Self {
+        NtxAccountIdPrefix((id.prefix().as_u64() >> 34) as u32)
+    }
+}
+
+impl TryFrom<NoteTag> for NtxAccountIdPrefix {
+    type Error = String;
+
+    fn try_from(tag: NoteTag) -> Result<Self, Self::Error> {
+        if tag.execution_mode() == NoteExecutionMode::Network && tag.is_single_target() {
+            return Ok(NtxAccountIdPrefix(tag.inner()));
+        }
+        Err("tag not meant for executing by a single account".into())
+    }
+}
+
+struct AccountEntry {
+    account: Account,
+    generation: usize,
+}
+
+struct CacheState {
+    generation: usize,
+    accounts: HashMap<NtxAccountIdPrefix, AccountEntry>,
+    ordering: BTreeMap<usize, NtxAccountIdPrefix>,
+}
+
+pub struct AccountCache {
+    capacity: usize,
+    state: Mutex<CacheState>,
+}
+
+impl AccountCache {
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            capacity,
+            state: std::sync::Mutex::new(CacheState {
+                generation: 0,
+                accounts: HashMap::with_capacity(capacity),
+                ordering: BTreeMap::new(),
+            }),
+        }
+    }
+
+    /// Insert or replace an entry.
+    pub fn put(&self, account: &Account) {
+        let account_id = NtxAccountIdPrefix::from(account.id());
+        let mut state = self.state.lock().unwrap();
+        state.generation += 1;
+        let current_generation = state.generation;
+
+        let removed_generation = if let Some(entry) = state.accounts.insert(
+            account_id,
+            AccountEntry {
+                generation: current_generation,
+                account: account.clone(),
+            },
+        ) {
+            Some(entry.generation)
+        } else {
+            None
+        };
+
+        if let Some(old) = removed_generation {
+            state.ordering.remove(&old);
+        }
+        state.ordering.insert(current_generation, account_id);
+
+        if state.accounts.len() > self.capacity {
```

Comment:
nit: tend to put explainer comments above even obvious operations like this to make it easier for people to scan through

Reply:
e.g.
```rust
// Remove oldest element from the cache if capacity full.
if state.accounts.len() > self.capacity {
...
```

Reply:
I was just pushing some changes which should address some of this, but tbh I think the `AccountCache` could still use a decent amount of polishing (left a `TODO`, maybe we should even replace it for a more robust/performant crate)